### PR TITLE
[jit] Support recursive ModuleList / Sequential

### PR DESCRIPTION
**Pull request by driazati**
Adds support for recursively compiling `nn.Sequential` and
`nn.ModuleList`. When either is used, it is converted to a
`jit._ConstModuleList` or `jit._ConstSequential` as necessary. Due to
this, we don't need to add it to `__constants__` since it's made
constant on demand.

This PR also moves the recursive script tests out to their own class
`TestRecursiveScript` (the added test is called `test_iterable_modules`)

Differential Revision: [D15611738](https://our.internmc.facebook.com/intern/diff/15611738/)

### CODE REVIEW

**Comment by eellison** [comment]
Can you include 'jit' in the name so that it is grepable and will be invoked when run_tests.py --jit is run

**Comment by driazati** [comment]
discussed in person, the test is still in `test_jit.py` so `run_test.py --jit` still runs it

**Comment by TheCodez** [comment]
@driazati does this also solve https://github.com/pytorch/pytorch/issues/20644 ?

**Comment by driazati** [comment]
@TheCodez it does in conjunction with #20939

File: torch/csrc/jit/script/python_sugared_value.cpp [hunk 1]
```diff
@@ -327,7 +326,7 @@ std::shared_ptr<SugaredValue> ModuleValue::attr(
         module_->register_module(field, submodule);
         auto v = module_->find_module(field);
         return std::make_shared<ModuleValue>(
-            m.graph()->insertGetAttr(self_, field), v, attr);
+            m.graph()->insertGetAttr(self_, field), v, result);
```
Comment (eellison) [inline torch/csrc/jit/script/python_sugared_value.cpp]: why was this changed ?

File: torch/jit/__init__.py [hunk 2]
```diff
@@ -940,11 +940,28 @@ def env(key):
...
+            module[i] = _create_constant_iterable_module(submodule)
+        else:
+            module[i] = _convert_to_script_module(submodule)
+
+    if isinstance(module, Sequential):
+        return _ConstSequential(module)
+    elif isinstance(module, ModuleList):
+        return _ConstModuleList(module)
+    else:
+        raise RuntimeError("Only nn.ModuleList and nn.Sequential can be made "
```
Comment (eellison) [inline torch/jit/__init__.py]: Seems like you would want to do the error checking that it's not a Sequential or ModuleList upfront so that you don't get some weird error when you try to enumerate through them

File: torch/jit/__init__.py [hunk 3]
```diff
@@ -953,6 +970,9 @@ def _try_compile_fn(fn):
         # Don't do anything for @ignore'd functions
         return None
 
+    if isinstance(fn, torch.nn.Module):
```
Comment (eellison) [inline torch/jit/__init__.py]: Could you add a comment here? It's i don't understand why we're returning None for both ignores and nn.Modules

File: torch/jit/__init__.py [hunk 4]
```diff
@@ -1651,6 +1682,10 @@ def _convert_to_script_module(mod, methods=None):
     `('forward',)`. Methods accessed in forward are scripted on demand if
     `_enable_recursive_script()` is used.
     """
+    if isinstance(mod, (ModuleList, Sequential)):
```
Comment (eellison) [inline torch/jit/__init__.py]: Why isn't this change not guarded by a `torch._C._jit_recursive_script` check

File: torch/csrc/jit/script/python_sugared_value.cpp (hunk 1 above)
Comment (driazati) [reply torch/csrc/jit/script/python_sugared_value.cpp]: It was wrong before and added the Python value, not the created scriptmodule

File: torch/jit/__init__.py (hunk 2 above)
Comment (driazati) [reply torch/jit/__init__.py]: This shouldn't be called on modules that aren't enumerable (it's only used in `_convert_to_script_module` and the type check is there)

File: torch/jit/__init__.py (hunk 4 above)
Comment (driazati) [reply torch/jit/__init__.py]: There's only 1 guard with that check in `script()` before it calls `_convert_to_script_module`, adding it everywhere would be cumbersome to delete later

File: torch/csrc/jit/script/python_sugared_value.cpp (hunk 1 above)
Comment (eellison) [reply torch/csrc/jit/script/python_sugared_value.cpp]: add test for this then ?

File: test/test_jit.py [hunk 5]
```diff
@@ -13475,6 +13395,125 @@ def foo(a):
...
+            super(TestRecursiveScript, self).run(result)
+
+    def checkModule(self, nn_module, args):
+        """
+        Check that a nn.Module's results in Script mode match eager and that it
+        can be exported
+        """
+        sm = torch.jit.script(nn_module)
+
+        eager_out = nn_module(*args)
```
Comment (eellison) [inline test/test_jit.py]: you should be freezing the RNG here

File: test/test_jit.py [hunk 6]
```diff
@@ -13475,6 +13395,125 @@ def foo(a):
...
+            __constants__ = ['x']
+
+            def __init__(self):
+                super(M, self).__init__()
+                self.other = Other(200)
+
+            def forward(self, t):
+                return self.other(t) * 2
+
+        sm = torch.jit.script(M())
```
Comment (eellison) [inline test/test_jit.py]: Why not use checkModule here

File: test/test_jit.py [hunk 7]
```diff
@@ -13475,6 +13395,125 @@ def foo(a):
...
+
+        self.assertExportImportModule(sm, (torch.ones(2, 2),))
```
Comment (eellison) [inline test/test_jit.py]: Same here

File: torch/jit/__init__.py [hunk 8]
```diff
@@ -1568,7 +1595,13 @@ def __init__(self, original, stubs):
                 elif isinstance(item, (Parameter, Module, Attribute)):
                     if isinstance(item, (ModuleList, Sequential)):
                         # These are in __constants__, so ignore them here
-                        continue
+
+                        if not torch._C._jit_recursive_script():
+                            # For recursive script, these are constantified after
```
Comment (eellison) [inline torch/jit/__init__.py]: Maybe add test for an example of when these are constantified and then attempted to be modified, if that applies

**Review by eellison — APPROVED** [review]
Seems good, couple more comments.

File: torch/jit/__init__.py [hunk 9]
```diff
@@ -962,7 +964,13 @@ def _make_strong_submodule(field, module, parent):
         # It's not a submodule, don't do anything
         return None
 
-    return _convert_to_script_module(module)
+    # Convert the module to a ScriptModule
+    new_strong_submodule = _convert_to_script_module(module)
+
+    # Install the ScriptModule on the python side
+    parent._modules._python_modules[field] = new_strong_submodule
```
Comment (eellison) [inline torch/jit/__init__.py]: Why did this change ?

Comment (driazati) [reply torch/jit/__init__.py]: It will have a `nn.Sequential` until it is used and compiled to a `_ConstSequential` and installs that as a submodule of the `._c` `ScriptModule`. This keeps the Python `ScriptModule` in sync by adding it there
